File: go/bob/bob.go
// Package bob should have a package comment that summarizes what it's about.
// https://golang.org/doc/effective_go.html#commentary
package bob

import (
	"strings"
	"unicode"
)

// Hey should have a comment documenting it.
func Hey(remark string) string {
	tr := strings.TrimSpace(remark)
	if tr == "" {
		return "Fine. Be that way!"
	}

	l := tr[len(tr)-1:]

	if isShouting(tr) && !isInt(tr) {
		if l == "?" {
			if !isLetter(tr[:len(tr)-1]) {
				return "Sure."
			}
			return "Calm down, I know what I'm doing!"
		}

		return "Whoa, chill out!"
	}

	if l == "?" {
		return "Sure."
	}

	return "Whatever."
}

func isShouting(line string) bool {
	for i := 0; i <= 1; i++ {
		if unicode.IsLower(rune(line[i])) {
			return false
		}
	}

	return true
}

func isInt(s string) bool {
	for _, c := range s {
		if c == rune(',') ||
			c == rune(' ') ||
			c == rune('?') ||
			c == rune('!') ||
			c == rune('\t') {
			continue
		}
		if !unicode.IsDigit(c) {
			return false
		}
	}
	return true
}

func isLetter(s string) bool {
	for _, r := range s {
		if r == rune(',') ||
			r == rune(' ') ||
			r == rune('?') ||
			r == rune('!') ||
			r == rune('\t') {
			continue
		}
		if !unicode.IsLetter(r) {
			return false
		}
	}
	return true
}

File: php/bob/bob.php
<?php

class Bob
{
    public function respondTo($m)
    {
        $m = trim($m);
        if ($m === '') return 'Fine. Be that way!';

        $l = substr($m, -1);

        if ($this->isShouting($m) && !$this->isNubmers($m)) {
            if ($l == "?") {
                if (!$this->isLetter($m)) {
                    return 'Sure.';
                }
                return 'Calm down, I know what I\'m doing!';
            }

                return "Whoa, chill out!";
        }

        if ($l == "?") {
            return 'Sure.';
        }

        return 'Whatever.';
    }

    public function isShouting($m) {
        for ($i=0; $i < strlen($m); $i++) {
            if ($this->isLegal($m[$i]) && ctype_lower($m[$i])) {
                return FALSE;
            }
        }

        return TRUE;
    }

    public function isNubmers($m) {
        for ($i=0; $i < strlen($m); $i++) {
            if ($this->isLegal($m[$i]) && !is_numeric($m[$i])) {
                return FALSE;
            }
        }

        return TRUE;
    }

    public function isLetter($m) {
        for ($i=0; $i < strlen($m); $i++) {
            if ($this->isLegal($m[$i]) && !ctype_alpha($m[$i])) {
                return FALSE;
            }
        }

        return TRUE;
    }

    private function isLegal(string $m): bool
    {
        if (in_array($m, [',', ' ', '?', '!', '\t'])) {
            return false;
        }

        return true;
    }
}
